PXC-3159: Killing the Donor or Connection lost during SST Process Leaves Joiner Hanging

venkatesh-prasad-v opened the pull request:
Problem
-------
When the donor node is killed while SST in progress, the joiner node
hung forever waiting for the SST to complete.

Analysis
--------
In general, if the donor node crashes while SST is in progress and if it
doesn't come back within the timeout decided by the
--sst-initial-timeout option, then SST script aborts with error 32
(Broken pipe) after the timeout is exceeded.

In such cases, Galera was not propagating the error to the above layers,
thereby causing other threads to wait for the state transfer to
complete.

Fix
---
Galera now handles the error by closing the group communication channels
and aborting the server.

Testing
----
Jenkins: https://pxc.cd.percona.com/view/PXC%208.0/job/pxc-8.0-param/72/testReport/

kamil-holubicki commented on sql/wsrep_xid.cc:
```diff
@@ -135,6 +136,10 @@ static bool get_SE_checkpoint(THD *, plugin_ref plugin, void *arg) {
     wsrep_uuid_print(&uuid, uuid_str, sizeof(uuid_str));
     WSREP_DEBUG("Read WSREPXid (InnoDB): %s:%lld", uuid_str,
                 (long long)wsrep_xid_seqno(xid));
+    DBUG_EXECUTE_IF("stop_after_reading_xid", {
```
I know that Jira ticket says about stopping here, but wouldn't it be enough to just stop when the donor thread is started in wsrep_sst.cc? If I understand the problem correctly we just want to kill the donor when the joiner is waiting for sst-meta-info, right?

venkatesh-prasad-v replied:
This is for the reason that `GCS_ACT_CCHANGE` is processed first on donor and then the `GCS_ACT_STATE_REQ` is processed.

It is during `GCS_ACT_CCHANGE`, the donor reads the xid from the SE and then delivers it to the joiner node so that it can decide whether to go for IST/SST, and then the replication thread on the joiner creates a `GCS_ACT_STATE_REQ` request and asks the donor to send the snapshot. On receiving the `GCS_ACT_STATE_REQ` on donor, it starts sst_donor_thread and proceeds with state transfer.

The issue with donor server going away during the `GCS_ACT_STATE_REQ` is handled properly on the joiner side and the server aborts with appropriate messages.

```
2020-07-15T06:28:17.575775Z 0 [ERROR] [MY-000000] [WSREP-SST] ******************* FATAL ERROR ********************** 
2020-07-15T06:28:17.581042Z 0 [ERROR] [MY-000000] [WSREP-SST] Possible timeout in receving first data from donor in gtid/keyring stage
2020-07-15T06:28:17.581079Z 0 [ERROR] [MY-000000] [WSREP-SST] Line 1108
2020-07-15T06:28:17.581092Z 0 [ERROR] [MY-000000] [WSREP-SST] ****************************************************** 
2020-07-15T06:28:17.581106Z 0 [ERROR] [MY-000000] [WSREP-SST] Cleanup after exit with status:32
2020-07-15T06:28:17.587344Z 0 [ERROR] [MY-000000] [WSREP] Process completed with error: wsrep_sst_xtrabackup-v2 --role 'joiner' --address '127.0.0.1:4002' --datadir '/home/venki/work/pxc/3159/80/bld/pxc-node/dn2/' --basedir '/home/venki/work/pxc/3159/80/bld/' --plugindir '/home/venki/work/pxc/3159/80/bld/lib/plugin/' --defaults-file '/home/venki/work/pxc/3159/80/bld/pxc-node/n2.cnf' --defaults-group-suffix '' --parent '20317' --mysqld-version '8.0.19-10'   '' : 32 (Broken pipe)
2020-07-15T06:28:17.587393Z 0 [ERROR] [MY-000000] [WSREP] Failed to read uuid:seqno from joiner script.
2020-07-15T06:28:17.587427Z 0 [ERROR] [MY-000000] [WSREP] SST script aborted with error 32 (Broken pipe)
2020-07-15T06:28:17.587574Z 3 [Note] [MY-000000] [Galera] Processing SST received
2020-07-15T06:28:17.587641Z 3 [Note] [MY-000000] [Galera] SST received: 00000000-0000-0000-0000-000000000000:-1
2020-07-15T06:28:17.587673Z 3 [System] [MY-000000] [WSREP] SST completed
2020-07-15T06:28:17.587763Z 2 [Note] [MY-000000] [Galera]  str_proto_ver_: 3 sst_seqno_: -1 cc_seqno: 2 req->ist_len(): 63
2020-07-15T06:28:17.587820Z 2 [ERROR] [MY-000000] [Galera] Application received wrong state: 
	Received: 00000000-0000-0000-0000-000000000000
	Required: 19be943e-c664-11ea-9be1-4f3dc6ddce72
2020-07-15T06:28:17.587838Z 2 [ERROR] [MY-000000] [Galera] Application state transfer failed. This is unrecoverable condition, restart required.
2020-07-15T06:28:17.587876Z 2 [Note] [MY-000000] [Galera] ReplicatorSMM::abort()

```
Since it is already being handled properly by shutting down the server, I didn't place the sync point in sst_donor_thread.

But however, processing of `GCS_ACT_CCHANGE` had some problems. So, I had to place the debug sync point in mysqld code(in wsrep_xid.cc).

kamil-holubicki reviewed (approved):
LGTM

Commit:
PXC-3159: Killing the Donor or Connection lost during SST Process Leaves Joiner
          Hanging

Problem
-------
When the donor node is killed while SST in progress, the joiner node
hung forever waiting for the SST to complete.

Analysis
--------
In general, if the donor node crashes while SST is in progress and if it
doesn't come back within the timeout decided by the
--sst-initial-timeout option, then SST script aborts with error 32
(Broken pipe) after the timeout is exceeded.

In such cases, Galera was not propagating the error to the above layers,
thereby causing other threads to wait for the state transfer to
complete.

Fix
---
Galera now handles the error by closing the group communication channels
and aborting the server.
